addons/knowledge/models/knowledge_article.py

```diff
@@ -0,0 +1,1423 @@
+        else:
+            return self.browse(my_articles_ids_keep) + other_article_res
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """ While creating records, automatically organize articles to be the
+        last of their parent children, unless a sequence is given.
+
+        Also allow to create private articles, using sudo if that case is
+        found. """
+        defaults = self.default_get(['article_member_ids', 'internal_permission', 'parent_id'])
+        vals_by_parent_id = {}
+        vals_as_sudo = []
+
+        for vals in vals_list:
+            can_sudo = False
+            # get values from vals or defaults
+            member_ids = vals.get('article_member_ids') or defaults.get('article_member_ids') or False
+            internal_permission = vals.get('internal_permission') or defaults.get('internal_permission') or False
+            parent_id = vals.get('parent_id') or defaults.get('parent_id') or False
+
+            # force write permission for workspace articles
+            if not parent_id and not internal_permission:
+                vals.update({'internal_permission': 'write',
+                             'parent_id': False,  # just be sure we don't grant privileges
+                })
+
+            # allow private articles creation
+            if not parent_id and internal_permission == 'none' and member_ids and len(member_ids) == 1:
+                self_member = member_ids[0][0] == Command.CREATE and member_ids[0][2].get('partner_id') == self.env.user.partner_id.id
+                if self_member:
+                    can_sudo = True
+
+            # if no sequence, parent will have to be checked
+            if not vals.get('sequence'):
+                vals_by_parent_id.setdefault(parent_id, []).append(vals)
+            vals_as_sudo.append(can_sudo)
+
+        # compute all maximum sequences / parent
+        max_sequence_by_parent = {}
+        if vals_by_parent_id:
+            parent_ids = list(vals_by_parent_id.keys())
+            try:
+                self.check_access_rights('write')
+                self.env['knowledge.article'].browse(parent_ids).check_access_rule('write')
+            except AccessError:
+                raise AccessError(_("You cannot create an article under articles on which you cannot write"))
+            max_sequence_by_parent = self._get_max_sequence_inside_parents(parent_ids)
+
+        # update sequences
+        for parent_id, article_vals in vals_by_parent_id.items():
+            current_sequence = 0
+            if parent_id in max_sequence_by_parent:
+                current_sequence = max_sequence_by_parent[parent_id] + 1
+
+            for vals in article_vals:
+                if 'sequence' in vals:
+                    current_sequence = vals.get('sequence')
+                else:
+                    vals['sequence'] = current_sequence
+                    current_sequence += 1
+
+        # sort by sudo / not sudo
```

Comment:
Why not splitting in two `if/else` ? You can't have both sudo and notsudo articles so the zip seems unnecessary :thinking:

Reply:
Actually you can, depending on given vals_list. To avoid using too much sudo.